### PR TITLE
Fix mypy typing errors attributed to `pytorch_lightning/demos/mnist_datamodule.py`

### PR DESCRIPTION
**Pull request by alro923**
## What does this PR do?

Fixes mypy typing errors in `pytorch_lightning/demos/mnist_datamodule.py` in https://github.com/Lightning-AI/lightning/issues/13445.

### Does your PR introduce any breaking changes? If yes, please list them.

None. just fixed typing errors.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃 → Yep! 😊

### CODE REVIEW

File: src/pytorch_lightning/demos/mnist_datamodule.py [hunk 1]
```diff
@@ -201,12 +199,12 @@ def setup(self, stage: Optional[str] = None) -> None:
...
         extra = dict(transform=self.default_transforms) if self.default_transforms else {}
         dataset: Dataset = MNIST(self.data_dir, train=True, download=False, **extra)
-        train_length = len(dataset)
+        train_length = len(dataset.data)
         self.dataset_train, self.dataset_val = random_split(dataset, [train_length - self.val_split, self.val_split])
 
     def train_dataloader(self) -> DataLoader:
         """MNIST train set removes a subset to use for validation."""
-        loader = DataLoader(
+        loader: DataLoader = DataLoader(
```
Comment (otaj) [inline src/pytorch_lightning/demos/mnist_datamodule.py:207]: I don't think it's required

File: src/pytorch_lightning/demos/mnist_datamodule.py [hunk 2]
```diff
@@ -218,7 +216,7 @@ def train_dataloader(self) -> DataLoader:
 
     def val_dataloader(self) -> DataLoader:
         """MNIST val set uses a subset of the training set for validation."""
-        loader = DataLoader(
+        loader: DataLoader = DataLoader(
```
Comment (otaj) [inline src/pytorch_lightning/demos/mnist_datamodule.py:219]: I don't think it's required

File: src/pytorch_lightning/demos/mnist_datamodule.py [hunk 3]
```diff
@@ -148,7 +148,7 @@ class MNISTDataModule(LightningDataModule):
     <...mnist_datamodule.MNISTDataModule object at ...>
     """
 
-    name = "mnist"
+    name: str = "mnist"
```
Comment (otaj) [inline src/pytorch_lightning/demos/mnist_datamodule.py:151]: I don't think this is required

File: src/pytorch_lightning/demos/mnist_datamodule.py [hunk 4]
```diff
@@ -99,7 +98,7 @@ def _download(self, data_folder: str) -> None:
             urllib.request.urlretrieve(url, fpath)
 
     @staticmethod
-    def _try_load(path_data: str, trials: int = 30, delta: float = 1.0) -> Tuple[Tensor, Tensor]:
+    def _try_load(path_data: str, trials: int = 30, delta: Union[int, float] = 1.0) -> Tuple[Tensor, Tensor]:
```
Comment (otaj) [inline src/pytorch_lightning/demos/mnist_datamodule.py:101]: I think this is absolutely fine staying just `float`

**Review by otaj — APPROVED** [review]
Left couple comments, but overall it looks good :) Thank you!

File: src/pytorch_lightning/demos/mnist_datamodule.py [hunk 5]
```diff
@@ -113,13 +112,14 @@ def _try_load(path_data: str, trials: int = 30, delta: float = 1.0) -> Tuple[Ten
                 time.sleep(delta * random.random())
             else:
                 break
+        assert res is not None, f"Failed to load data from {path_data}"
```
Comment (rohitgr7) [inline src/pytorch_lightning/demos/mnist_datamodule.py:115]:
```suggestion
        assert res is not None
```

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
@otaj @rohitgr7 I addressed your reviews

File: src/pytorch_lightning/demos/mnist_datamodule.py [hunk 6]
```diff
@@ -201,7 +199,7 @@ def setup(self, stage: Optional[str] = None) -> None:
         """Split the train and valid dataset."""
         extra = dict(transform=self.default_transforms) if self.default_transforms else {}
         dataset: Dataset = MNIST(self.data_dir, train=True, download=False, **extra)
-        train_length = len(dataset)
+        train_length = len(dataset.data)
```
Comment (rohitgr7) [inline src/pytorch_lightning/demos/mnist_datamodule.py:202]:
```suggestion
        train_length = len(dataset)
```
I don't think this is required.

**Review by rohitgr7 — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
